File: src/main/scala/eschool/sites/model/Site.scala
package eschool.sites.model

import scala.collection.JavaConverters._
import javax.jdo.annotations._
import eschool.users.model.User
import jdo.Id

@Uniques(Array(
  new Unique(members=Array("owner", "name")), 
  new Unique(members=Array("owner", "ident"))))
class Site extends Id[Long] {
  private[this] var _owner: User = _
  private[this] var _name: String = _
  private[this] var _ident: String = _
  @Join
  private[this] var _children: java.util.List[Page] = _
  
  def this(owner: User, name: String, ident: String) = {
    this()
    owner_=(owner)
    name_=(name)
    ident_=(ident)
  }

  def owner: User = _owner
  def owner_=(theOwner: User) { _owner = theOwner }
  
  def name: String = _name
  def name_=(theName: String) { _name = theName }
  
  def ident: String = _ident
  def ident_=(theIdent: String) { _ident = theIdent }
  
  def children: List[Page] = {
    if (_children == null) List[Page]()
    else _children.asScala.toList
  }
  def children_=(theChildren: List[Page]) {
    theChildren foreach ((page: Page) => {
      page.parentSite = this
    })
    _children = theChildren.asJava
  } 
}

object Site {
  
}

File: src/main/java/eschool/sites/model/Site.java
package eschool.sites.model;

import java.util.Map;

import javax.jdo.annotations.*;

import eschool.users.model.User;

@PersistenceCapable
@Uniques({@Unique(members={"owner", "name"}), @Unique(members={"owner", "ident"})})
public class Site {
	@PrimaryKey
	@Persistent(valueStrategy=IdGeneratorStrategy.INCREMENT)
	private long id;
	private User owner;
	private String name;
	private String ident;
	private Map<String, Page> children;
	
	public Site() {}
	
	public Site(User owner, String name, String ident) {
		this.owner = owner;
		this.name = name;
		this.ident = ident;
	}
	
	public void setChildren(Map<String, Page> children) {
		for (String ident : children.keySet()) {
			Page p = children.get(ident);
			p.setIdent(ident);
			p.setParentSite(this);
		}
		this.children = children;
	}
	
	public Map<String, Page> getChildren() {
		return this.children;
	}

	public User getOwner() {
		return owner;
	}

	public void setOwner(User owner) {
		this.owner = owner;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public String getIdent() {
		return ident;
	}

	public void setIdent(String ident) {
		this.ident = ident;
	}

	public long getId() {
		return id;
	}
}
